### ISSUE TITLE
Index out of bounds: the len is 26 but the index is 30

### ISSUE THREAD

**Issue by qarmin** [typst/typst#4573]
### Description

File content(at the bottom should be attached raw, not formatted file - github removes some non-printable characters, so copying from here may not work):
```
#let apa-bibliography = (() => {
let dict-from-entries = entries => {
	for (key, value) in entries {
let n(x, default, f) = if x == none { default } else  f(x) }
	}
}
let months = (
)
let parse-date(x) = {
(works, include-all: true) => {
	let use-workkey) = [#used-works.update(x => x + (key,))]
```

command
```
timeout -v 100 typst compile TEST___FILE.typ
```

cause this
```
thread 'main' panicked at crates/typst-syntax/src/parser.rs:838:7:
index out of bounds: the len is 26 but the index is 30
stack backtrace:
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

##### Automatic Fuzzer note, output status "Some(101)", output signal "None"
```

[compressed.zip](https://github.com/user-attachments/files/16265972/compressed.zip)

### Operating system

Linux

### Typst version

- [X] I am using the latest version of Typst

**Comment by bluebear94** [comment]
Seems to be caused by #3463. If I disable the memoization in `expr_with_paren`, then the panic doesn’t occur.

**Comment by bluebear94** [comment]
I think this happens because the parser needs to parse a parenthesized expression at the beginning of the line containing `(works, include-all: true) => {` in two different situations, with two different lists of accumulated syntax nodes in `p.nodes`. Not sure how to fix this; we could memoize by `(p.current_start, p.nodes.len())` instead of by `p.current_start` only, but that might hide a deeper issue.

**Comment by wrzian** [comment]
Good news! This was resolved by the parser refactor (#5310) and should ship with the next version update. In particular it was commit 5 (16cc7eb), which refactored the memoization system, and mainly this comment in `restore_memo_or_checkpoint`:

```rust
// It's important that we don't truncate the nodes vector since
// it may have grown or shrunk (due to other memoization or
// error reporting) since we made this checkpoint.
```

In this case, the number of errors increases between memoization steps, and the old version incorrectly truncated the nodes vector, leading to an out-of-bounds access plus panic when we report the final unclosed curly-brace.

The most minimal reproduction I could make is:
```typ
#{
  (
    if x {} else {}
    { () = 2
  ) = 3
}
```

Let's go over each of the parts:

- Setup
    - We use a code-block just so we can write an assignment expression
    - We write an assignment expression with parentheses on the left side, which will cause a mispredict in the paren expression parser
    - Inside the parens, we use the if-statement to create a case that has fewer errors when parsed as a parenthesized expression vs. as the lhs of an assignment. (keywords aren't allowed at all in destructuring assignments, so it errors early)
    - We start an unclosed curly-brace expression
    - After the open curly-brace we add another assignment expression so the paren parser will mispredict, causing a reload of the nodes vector.
- Parsing
    - In the first pass, the parser creates the unclosed-delimiter error message with no problems
    - But when the parser reaches the outer equals sign, `) = 3`, it reparses the inside as a destructuring pattern.
    - This time, the parser generates more errors (so the node length is longer) and the index of the open curly-brace is further in the nodes vector
    - When we reach the inner equals, `() = 2`, we load the value for it from the memo and the old version would incorrectly truncate the nodes vector when loading the parsed state
    - When we try to create the unclosed-delimiter error, we access the opening delimiter by it's new, longer, index
    - But this access is out-of-bounds because the nodes vector was truncated
    - So Rust panics

What a lovely bug!
:boot:
:bug:

**Comment by mkorje** [comment]
I think this can be closed now.
